src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -807,63 +807,206 @@ ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query, RangeTblEntry *inse
+		List *insertTargetEntryColumnList = pull_var_clause_default((Node *) targetEntry);
+		int insertTargetEntryColumnCount = list_length(insertTargetEntryColumnList);
+		Var *insertVar = NULL;
+		AttrNumber originalAttrNo = InvalidAttrNumber;
+		TargetEntry *subqeryTargetEntry = NULL;
 
-		if (IsA(targetEntry->expr, Var))
+		/*
+		 * We only consider target entries that include a single column. Note that this
+		 * is slightly different than directly checking the whether the targetEntry->expr
+		 * is a var since the var could be wrapped into an implicit/explicit casting.
+		 */
+		if (insertTargetEntryColumnCount != 1)
 		{
-			Var *insertVar = (Var *) targetEntry->expr;
-			AttrNumber originalAttrNo = get_attnum(insertRelationId,
-												   targetEntry->resname);
-			TargetEntry *subqeryTargetEntry = NULL;
+			continue;
+		}
 
-			if (originalAttrNo != insertPartitionColumn->varattno)
-			{
-				continue;
-			}
+		insertVar = (Var *) linitial(insertTargetEntryColumnList);
+		originalAttrNo = get_attnum(insertRelationId,
+									targetEntry->resname);
+
+		/* skip non-partition column on the target table */
+		if (originalAttrNo != insertPartitionColumn->varattno)
+		{
+			continue;
+		}
+
+		/* INSERT query includes the partition column */
+		targetTableHasPartitionColumn = true;
+
+		subqeryTargetEntry = list_nth(subquery->targetList,
+									  insertVar->varattno - 1);
+
+		/*
+		 * Corresponding (i.e., in the same ordinal position as the target table's
+		 * partition column) select target entry does not directly belong a table.
+		 * Evaluate its expression type and error out properly.
+		 */
+		if (subqeryTargetEntry->resorigtbl == InvalidOid)
+		{
+			char *exprTypeDefault = "an expression that is not a plain column";
+			char *errorDetailTemplate = "Subquery contains %s in the "
+										"same position of the target table's "
+										"partition column.";
+
+			Expr *selectTargetExpr = subqeryTargetEntry->expr;
+			NodeTag exprType = selectTargetExpr->type;
+			char *errorDetail = "";
+			char *exprTypeStr = "";
 
-			subqeryTargetEntry = list_nth(subquery->targetList,
-										  insertVar->varattno - 1);
 
-			if (!IsA(subqeryTargetEntry->expr, Var))
+			switch (exprType)
 			{
-				partitionColumnsMatch = false;
+				case T_Const:
+				{
+					exprTypeStr = "a constant value";
+				}
 				break;
-			}
 
-			/*
-			 * Reference tables doesn't have a partition column, thus partition columns
-			 * cannot match at all.
-			 */
-			if (PartitionMethod(subqeryTargetEntry->resorigtbl) == DISTRIBUTE_BY_NONE)
-			{
-				partitionColumnsMatch = false;
+				case T_OpExpr:
+				{
+					exprTypeStr = "an operator";
+				}
 				break;
-			}
 
-			if (!IsPartitionColumnRecursive(subqeryTargetEntry->expr, subquery))
-			{
-				partitionColumnsMatch = false;
+				case T_FuncExpr:
+				{
+					FuncExpr *subqueryFunctionExpr =
+						(FuncExpr *) subqeryTargetEntry->expr;
```

Comment:
You already have `selectTargetExpr`, so just cast that rather than rewrite the `->expr`.